libr/asm/arch/pyc/pyc_dis.c

```diff
@@ -0,0 +1,211 @@
+			if (arg != NULL) {
+				r_strbuf_appendf (&opstruct->buf_asm, "%20s", arg);
+			}
+		} else if (ops->bits == 8) {
+			i += 1;
+		}
+
+		return i;
+	}
+	return 0;
+}
+
+static const char *parse_arg(pyc_opcode_object *op, ut32 oparg, RList *names, RList *consts, RList *varnames, RList *interned_table, RList *freevars, RList *cellvars, RList *opcode_arg_fmt) {
+	pyc_object *t = NULL;
+	const char *arg = NULL;
+	pyc_code_object *tmp_cobj;
+	pyc_arg_fmt *fmt;
+	RListIter *i = NULL;
+
+	// version-specific formatter for certain opcodes
+	r_list_foreach (opcode_arg_fmt, i, fmt)
+		if (!strcmp (fmt->op_name, op->op_name)) {
+			return fmt->formatter (oparg);
+		}
+
+	if (op->type & HASCONST) {
+		t = (pyc_object *)r_list_get_n (consts, oparg);
+		if (t == NULL) {
+			return NULL;
+		}
+		switch (t->type) {
+		case TYPE_CODE_v0:
+		case TYPE_CODE_v1:
+			tmp_cobj = t->data;
+			arg = r_str_newf ("CodeObject(%s) from %s", tmp_cobj->name->data, tmp_cobj->filename->data);
+			break;
+		case TYPE_TUPLE:
+		case TYPE_SET:
+		case TYPE_FROZENSET:
+		case TYPE_LIST:
+		case TYPE_SMALL_TUPLE:
+			arg = generic_array_obj_to_string (t->data);
+			break;
+		case TYPE_STRING:
+		case TYPE_INTERNED:
+		case TYPE_STRINGREF:
+			arg = r_str_newf ("'%s'", t->data);
+		default:
+			arg = t->data;
+		}
+	}
+	if (op->type & HASNAME) {
+		t = (pyc_object *)r_list_get_n (names, oparg);
+		if (t == NULL) {
+			return NULL;
+		}
+		arg = t->data;
+	}
+	if ((op->type & HASJREL) || (op->type & HASJABS)) {
+		arg = r_str_newf ("%u", oparg);
+	}
+	if (op->type & HASLOCAL) {
+		t = (pyc_object *)r_list_get_n (varnames, oparg);
+		if (!t)
+			return NULL;
+		arg = t->data;
+	}
+	if (op->type & HASCOMPARE) {
+		arg = cmp_op[oparg];
+	}
+	if (op->type & HASFREE) {
+		if (!cellvars || !freevars) {
+			arg = r_str_newf ("%u", oparg);
+			return arg;
+		}
+
+		if (oparg < r_list_length (cellvars)) {
+			t = (pyc_object *)r_list_get_n (cellvars, oparg);
+		} else if ((oparg - r_list_length (cellvars)) < r_list_length (freevars)) {
+			t = (pyc_object *)r_list_get_n (freevars, oparg);
+		} else {
+			arg = r_str_newf ("%u", oparg);
+			return arg;
+		}
+		if (!t) {
+			return NULL;
+		}
+
+		arg = t->data;
+	}
+	if (op->type & HASNARGS) {
+		arg = r_str_newf ("%u", oparg);
+	}
+	if (op->type & HASVARGS) {
+		arg = r_str_newf ("%u", oparg);
+	}
+
+	return arg;
+}
+
+/* for debugging purpose */
+void dump(RList *l) {
+	RListIter *it;
+	pyc_object *e = NULL;
+
+	r_list_foreach (l, it, e) {
+		if (e->type == TYPE_TUPLE) {
+			eprintf ("[TYPE_TUPLE] %s\n", generic_array_obj_to_string (e->data));
+			return;
+		}
+		eprintf ("[DATA] %s\n", (char *)e->data);
+	}
+}
+
+char *generic_array_obj_to_string(RList *l) {
+	RListIter *iter = NULL;
+	pyc_object *e = NULL;
+	ut32 size = 256, used = 0;
+	char *r = NULL, *buf = NULL;
+
+	// add good enough space
+	buf = (char *)calloc (size + 10, 1);
+	r_list_foreach (l, iter, e) {
+		while (!(strlen (e->data) < size)) {
```

Comment:
seems like strlen(e->data) is done 3 times in this loop. just hold the value once and reuse the variable